src/main/java/games/strategy/util/memento/PropertyBagMementoImporter.java

```
@@ -0,0 +1,114 @@
+    return new MementoImportException(String.format("memento has wrong type: expected '%s' but was '%s'",
+        PropertyBagMemento.class.getName(), memento.getClass().getName()));
+  }
+
+  private static MementoImportException newUnsupportedSchemaIdException(final String schemaId) {
+    return new MementoImportException(String.format("schema ID '%s' is unsupported", schemaId));
+  }
+
+  private static MementoImportException newUnsupportedSchemaVersionException(final long schemaVersion) {
+    return new MementoImportException(String.format("schema version %d is unsupported", schemaVersion));
+  }
+
+  /**
+   * A memento import handler.
+   *
+   * @param <T> The type of the memento originator.
+   */
+  @FunctionalInterface
+  public interface Handler<T> {
+    /**
+     * Imports the specified originator properties and creates a new originator instance.
+     *
+     * @param propertiesByName The collection of originator properties; must not be {@code null}. The key
+     *        is the property name. The value is the property value.
+     *
+     * @return A new originator instance created from the properties; never {@code null}.
+     *
+     * @throws MementoImportException If an error occurs while importing the originator properties.
+     */
+    T importProperties(Map<String, Object> propertiesByName) throws MementoImportException;
+  }
+
+  /**
+   * Supplies memento import handlers for each supported schema version.
+   *
+   * @param <T> The type of the memento originator.
+   */
+  @FunctionalInterface
+  public interface HandlerSupplier<T> {
+    /**
+     * Gets the memento import handler for the specified schema version.
+     *
+     * @param schemaVersion The memento schema version to import.
+     *
+     * @return The memento import handler for the specified schema version or empty if no handler is available; never
+     *         {@code null}.
+     */
+    Optional<Handler<T>> getHandler(long schemaVersion);
```

The one place I can see this is called, it appears required:
> +    final Handler<T> handler = handlerSupplier.getHandler(memento.getSchemaVersion())
 +        .orElseThrow(() -> newUnsupportedSchemaVersionException(memento.getSchemaVersion()));

Why return optional at all? When we set the handler, if we would set it to null/empty, we could fail sooner at that time rather than waiting for someone to read the null/empty property.

Recommend dropping the "Optional" from the API.